misc: Increase Clang support to v19; Drop GCC v10 support

BobbyRBruce opened the pull request:
- Increases the supported Clang version to 19.
- Removes support for GCC v10, making the oldest supported version v11.
- Converts compiler version errors to warnings, thus allowing use of unsupported compilers with just a warning thrown.

BobbyRBruce committed:
tests,misc: Add support for Clang v19

Ubuntu 24.04 is the latest LTS release with an APT package manager
which enables installation of up to Clang v19.

BobbyRBruce commented:
I would've liked to increase to v20 but v19 is the highest you can get on Ubuntu 24.04 with out-of-the-box APT. Supporting the latest Ubuntu LTS release APT package library is normally what we aim for.

abmerop commented:
Do we have a list of the _minimum_ compiler versions gem5 supports?  If we do, I could not find it.  I'd love to use some features from newer compilers but it requires gcc 12+ and clang 15+ for example.

BobbyRBruce committed:
scons: Update supported clang/gcc error/warnings

1) Ensure Scons is working with the right min and max version for Clang
   and GCC.
2) Replace the Errors with Warnings: While it's not recommended to
   compile gem5 with a compiler not supported by the project; they
   are permitted to do so if they wish. A warning is more apt in this
   case.

BobbyRBruce committed:
scons: Rm Clang Scons branch for unsupported Clang versions

BobbyRBruce commented:
I had a go at removing GCC 10 as it's quite old now, but gcn-gpu container requires gcc 10 so I'll keep it for now.

BobbyRBruce commented:
I've decided to drop v10 support. In addition to this change being added I also changed the errors thrown is using unsupported compiler versions to warnings. I think this is more apt (an older/newer compiler may well work,  it's ust that we don't support them)

BobbyRBruce committed:
misc,scons: Increase min GCC version support to v11

BobbyRBruce committed:
scons: Rm erroneous escape chars in f-strs

BobbyRBruce committed:
scons: Fix 'warn' -> 'warning'

BobbyRBruce committed:
scons: Fix incorrect clang version warning condition
